### PR TITLE
feat: add base protocols for Resource and ResourceSequence

### PR DESCRIPTION
**Pull request by tdstein**

### CODE REVIEW

File: src/posit/connect/resources.py [hunk 1]
```diff
@@ -77,7 +88,26 @@ def update(self, **attributes):  # type: ignore[reportIncompatibleMethodOverride
         super().update(**result)
 
 
-class _ResourceSequence(Sequence):
+T = TypeVar("T", bound=Resource)
+
+
+class ResourceSequence(Protocol[T]):
```
Comment (schloerke) [inline src/posit/connect/resources.py:94]: Thank you

**Review by schloerke — APPROVED** [review]

File: src/posit/connect/resources.py [hunk 2]
```diff
@@ -62,7 +69,11 @@ def __init__(self, ctx: Context, path: str, /, **attributes):
         self._path = path
 
 
-class _Resource(dict):
+class Resource(Protocol):
+    def __getitem__(self, key: Hashable) -> Any: ...
+
+
+class _Resource(dict, Resource):
```
Comment (schloerke) [inline src/posit/connect/resources.py:76]: This class is no longer needed?

File: src/posit/connect/resources.py [hunk 3]
```diff
@@ -62,7 +69,11 @@ def __init__(self, ctx: Context, path: str, /, **attributes):
         self._path = path
 
 
-class _Resource(dict):
+class Resource(Protocol):
+    def __getitem__(self, key: Hashable) -> Any: ...
```
Comment (schloerke) [inline src/posit/connect/resources.py:73]:
```suggestion
    def __getitem__(self, key: Hashable, /) -> Any: ...
```

File: src/posit/connect/resources.py (hunk 2 above)
Comment (tdstein) [reply src/posit/connect/resources.py:76]: Correct. The BaseResource class should go away. But that likely requires finishing the migration to the new implicit subtyping pattern. Which I am planning to complete.

**Pull request merged**